File: app/src/main/java/fr/gstraymond/biz/SearchOptions.java
package fr.gstraymond.biz;

import android.os.Parcel;
import android.os.Parcelable;
import android.text.TextUtils;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import fr.gstraymond.android.CustomParcelable;

public class SearchOptions extends CustomParcelable {

    public static final String QUERY_ALL = "*";

    private String query = QUERY_ALL;
    private boolean append = false;
    private boolean random = false;
    private boolean addToHistory = true;
    private int from = 0;
    private int size = 30;
    private Facets facets = new Facets();
    private Map<String, Integer> facetSize = new HashMap<String, Integer>();


    public static final Parcelable.Creator<SearchOptions> CREATOR = new Parcelable.Creator<SearchOptions>() {
        @Override
        public SearchOptions createFromParcel(Parcel source) {
            return new SearchOptions(source);
        }

        @Override
        public SearchOptions[] newArray(int size) {
            return new SearchOptions[size];
        }
    };

    public SearchOptions(Parcel source) {
        query = source.readString();
        append = source.readInt() == 0;
        random = source.readInt() == 0;
        from = source.readInt();
        size = source.readInt();
        facets = readFacets(source);
        // facetSize : pas de persistence de la taille des facettes
        // addToHistory : pas de persistence de l'ajout à l'historique
    }

    public SearchOptions() {
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(query);
        dest.writeInt(append ? 0 : 1);
        dest.writeInt(random ? 0 : 1);
        dest.writeInt(from);
        dest.writeInt(size);
        writeFacets(dest, facets);
        // facetSize : pas de persistence de la taille des facettes
        // addToHistory : pas de persistence de l'ajout à l'historique
    }

    public String getQuery() {
        return query;
    }

    public SearchOptions setQuery(String query) {
        if (TextUtils.isEmpty(query)) {
            this.query = QUERY_ALL;
        } else {
            this.query = query;
        }
        return this;
    }

    public boolean isAppend() {
        return append;
    }

    public SearchOptions setAppend(boolean append) {
        this.append = append;
        return this;
    }

    public boolean isRandom() {
        return random;
    }

    public SearchOptions setRandom(boolean random) {
        this.random = random;
        return this;
    }

    public Integer getFrom() {
        return from;
    }

    public SearchOptions setFrom(Integer from) {
        this.from = from;
        return this;
    }

    public Integer getSize() {
        return size;
    }

    public SearchOptions setSize(Integer size) {
        this.size = size;
        return this;
    }

    public SearchOptions addFacet(String facet, String term) {
        if (facets.containsKey(facet)) {
            facets.get(facet).add(term);
        } else {
            List<String> terms = new ArrayList<String>();
            terms.add(term);
            facets.put(facet, terms);
        }
        return this;
    }

    public SearchOptions removeFacet(String facet, String term) {
        if (facets.containsKey(facet)) {
            facets.get(facet).remove(term);

            if (facets.get(facet).size() == 0) {
                facets.remove(facet);
            }
        }
        return this;
    }

    public SearchOptions addFacetSize(String facet) {
        if (facetSize.containsKey(facet)) {
            Integer size = facetSize.get(facet);
            facetSize.put(facet, size + 10);
        } else {
            facetSize.put(facet, 20);
        }
        return this;
    }

    public Facets getFacets() {
        return facets;
    }

    public SearchOptions setFacets(Facets facets) {
        this.facets = facets;
        return this;
    }

    public boolean isAddToHistory() {
        return addToHistory;
    }

    public SearchOptions setAddToHistory(boolean addToHistory) {
        this.addToHistory = addToHistory;
        return this;
    }

    public Map<String, Integer> getFacetSize() {
        return facetSize;
    }

    public void setFacetSize(Map<String, Integer> facetSize) {
        this.facetSize = facetSize;
    }

    @Override
    public String toString() {
        return String.format(
                "searchOptions:[query:%s, append:%s, random:%s, addToHistory:%s, from:%s, size:%s, facets:%s, facetSize:%S]",
                query,
                append,
                random,
                addToHistory,
                from,
                size,
                facets,
                facetSize);
    }
}

File: app/src/main/kotlin/fr/gstraymond/biz/SearchOptions.kt
package fr.gstraymond.biz

import android.os.Parcel
import android.os.Parcelable
import android.text.TextUtils
import fr.gstraymond.utils.FacetParcelableUtils.readFacets
import fr.gstraymond.utils.FacetParcelableUtils.writeFacets

data class SearchOptions(var query: String = QUERY_ALL,
                         var append: Boolean = false,
                         var random: Boolean = false,
                         var addToHistory: Boolean = true,
                         var from: Int = 0,
                         var size: Int = 30,
                         var facets: Map<String, List<String>> = hashMapOf(),
                         var facetSize: Map<String, Int> = hashMapOf(),
                         var deckId: String? = null,
                         var sort: String? = null,
                         var addToSideboard: Boolean = false) : Parcelable {

    constructor(source: Parcel) : this(
            query = source.readString(),
            append = source.readInt() == 0,
            random = source.readInt() == 0,
            from = source.readInt(),
            size = source.readInt(),
            facets = readFacets(source),
            deckId = source.readString(),
            sort = source.readString(),
            addToSideboard = source.readInt() == 0
            // facetSize : pas de persistence de la taille des facettes
            // addToHistory : pas de persistence de l'ajout à l'historique
            // fromOkGoogle : pas de persistence
    )

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(query)
        dest.writeInt(if (append) 0 else 1)
        dest.writeInt(if (random) 0 else 1)
        dest.writeInt(from)
        dest.writeInt(size)
        writeFacets(dest, facets)
        dest.writeString(deckId)
        dest.writeString(sort)
        dest.writeInt(if (addToSideboard) 0 else 1)
        // facetSize : pas de persistence de la taille des facettes
        // addToHistory : pas de persistence de l'ajout à l'historique
        // fromOkGoogle : pas de persistence
    }

    fun updateQuery(query: String) = apply {
        this.query = when {
            TextUtils.isEmpty(query) -> QUERY_ALL
            else -> query
        }
    }

    fun updateAppend(append: Boolean) = apply {
        this.append = append
    }

    fun updateFrom(from: Int) = apply {
        this.from = from
    }

    fun addFacet(facet: String, term: String) = apply {
        val terms = facets.getOrElse(facet) { listOf() } + term
        facets += facet to terms
    }

    fun removeFacet(facet: String, term: String) = apply {
        val terms = facets.getOrElse(facet) { listOf() } - term
        facets = when {
            terms.isEmpty() -> facets.filterKeys { it != facet }
            else -> facets + (facet to terms)
        }
    }

    fun addFacetSize(facet: String) = apply {
        val size = facetSize.getOrElse(facet) { 10 } + 10
        facetSize += facet to size
    }

    fun updateFacets(facets: Map<String, List<String>>) = apply {
        this.facets = facets
    }

    fun updateAddToHistory(addToHistory: Boolean) = apply {
        this.addToHistory = addToHistory
    }

    fun updateSize(size: Int) = apply {
        this.size = size
    }

    fun updateSort(sort: String?) = apply {
        this.sort = sort
    }

    companion object {
        val QUERY_ALL = "*"

        fun START_SEARCH_OPTIONS() = SearchOptions(
                size = 1,
                sort = "publications.editionReleaseDate:desc,publications.collectorNumber:desc",
                addToHistory = false)

        @JvmField
        val CREATOR = object : Parcelable.Creator<SearchOptions> {
            override fun createFromParcel(source: Parcel) = SearchOptions(source)

            override fun newArray(size: Int) = arrayOfNulls<SearchOptions>(size)
        }
    }
}
